From f3f544f99884415af888e6298655bd2459065168 Mon Sep 17 00:00:00 2001
From: raosrini <srinidhi.rao@intel.com>
Date: Mon, 5 Oct 2015 14:26:40 +0530
Subject: [PATCH] EM/bq24192_charger: WA to fix the Status reg updation during
 reboot

In BQ CHG probe, DPDM bit is force set to true inorder to force a
DPDM detection and dectect the charger cable if it was inserted
prior to boot up.
However, when we force trigger the DPDM detection from probe,
as per the Vendor data sheet, it requires a delay of 50ms
to complete the DPDM detection.

This patch adds a flag 'dpdm_trig' which will be set during
probe when DPDM detection is set. In the IRQ thread function,
a sleep of 50ms is added if the dpdm_trig flag is set. The
flag will be susequently reset to false state.

This patch has been ported from L-Mr1-CHT_Stable branch.

Change-Id: I9eff42dfdfa7bd7e59fb9b36b6c2b5ef7abd3937
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-1175
Signed-off-by: raosrini <srinidhi.rao@intel.com>
---
 drivers/power/bq24192_charger.c |   23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

Index: b/drivers/power/bq24192_charger.c
===================================================================
--- a/drivers/power/bq24192_charger.c	2016-03-08 11:04:29.862545363 -0800
+++ b/drivers/power/bq24192_charger.c	2016-03-08 11:09:40.025852667 -0800
@@ -48,7 +48,6 @@
 
 #include <asm/intel_em_config.h>
 
-
 #define DRV_NAME "bq24192_charger"
 #define DEV_NAME "bq24192"
 #define VBUS_CTRL_CDEV_NAME	"vbus_control"
@@ -298,6 +297,7 @@
 	bool online;
 	bool present;
 	bool chrg_usb_compliance;
+	bool dpdm_trig;
 };
 
 enum vbus_states {
@@ -1781,6 +1781,26 @@
 						    struct bq24192_chip,
 						    irq_wrkr.work);
 
+	/*
+	 * VBUS status updation in the Status Register of BQ24297 chip
+	 * is bit delayed, which is causing a wrong VBUS status
+	 * notification to USB and power_supply. So a delay of 10 milli
+	 * seconds is added before reading Status Register.
+	 * However, during the booup, in probe, we force set the DPDM
+	 * detection in order to get the cable status if any charger
+	 * cable is connected prior to the device boot up. Hence if
+	 * DPDM is force triggered, as per the Data sheet, the DPDM
+	 * detect operation will take 50ms to complete. Hence imparting
+	 * Sleep for 50ms if dpdm is force triggered.
+	 */
+	if (chip->chip_type == BQ24297) {
+		if (chip->dpdm_trig) {
+			msleep(50);
+			chip->dpdm_trig = false;
+		} else
+			msleep(10);
+	}
+
 	/**
 	 * check the bq24192 status/fault registers to see what is the
 	 * source of the interrupt
@@ -2496,6 +2516,7 @@
 		 * charger cable connected before hand and will trigger
 		 * an interrupt accordingly.
 		 */
+		chip->dpdm_trig = true;
 		ret = bq24192_reg_read_modify(chip->client,
 			BQ24192_MISC_OP_CNTL_REG, MISC_OP_CNTL_DPDM_EN, true);
 		if (ret < 0)
